### PR TITLE
Add multi account provisioning

### LINKED ISSUE #4459: Multi-account provisioning
### Feature Request

As an admin I might not just have one email server configuration for my users but their server configuration might depend on the email address. I can provision their email address through LDAP, but in Nextcloud Mail there is currently only the option for one IMAP/SMTP configuration. There should be multiple of them and each is only active for a certain domain. If the user logs in, the app will check for a matching configuration and then us that one.

#### Summary

Allow multiple provisioned configs and pick config based on user email.

### PR DESCRIPTION
**Pull request by miaulalala**
Rework this feature to use a table instead of the JSON that's saved in the APPVALUE parameter.

Fixes #4459

### CODE REVIEW

File: lib/Db/ProvisioningMapper.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		parent::__construct($db, 'mail_provisionings');
+	}
+
+	public function getActive() : ?array {
+		$qb = $this->db->getQueryBuilder();
+		$qb = $qb->select('*')
+				->from($this->getTableName())
+				->where($qb->expr()->eq('active', $qb->createNamedParameter('true', IQueryBuilder::PARAM_BOOL)));
+		$active = $qb->execute();
+		$result = $active->fetchAll();
```
Comment (ChristophWurst) [inline lib/Db/ProvisioningMapper.php:46]: ``\OCP\AppFramework\Db\QBMapper::findOneQuery`` is your friend :)

File: lib/Db/ProvisioningMapper.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		$active->closeCursor();
+		return $result;
+	}
+
+	public function getAll() : ?array {
+		$qb = $this->db->getQueryBuilder();
+		$qb = $qb->select('*')
+				->from($this->getTableName());
+		$all = $qb->execute();
+		$result = $all->fetchAll();
```
Comment (ChristophWurst) [inline lib/Db/ProvisioningMapper.php:56]: ``\OCP\AppFramework\Db\QBMapper::findEntities``

File: lib/Db/ProvisioningMapper.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		return $result;
+	}
+
+	public function getWildcardConfig() : ?array {
+		$qb = $this->db->getQueryBuilder();
+		$qb = $qb->select('*')
+				->from($this->getTableName())
+				->where($qb->expr()->eq('provisioning_domain', $qb->createNamedParameter('*', IQueryBuilder::PARAM_BOOL)));
+		$wildcard = $qb->execute();
+		$result = $wildcard->fetch();
```
Comment (ChristophWurst) [inline lib/Db/ProvisioningMapper.php:67]: ``\OCP\AppFramework\Db\QBMapper::findEntities``

File: lib/Model/IMAPMessage.php [hunk 4]
```diff
@@ -724,14 +724,17 @@ public function toDbMessage(int $mailboxId, MailAccount $account): Message {
...
 			Horde_Imap_Client::FLAG_ANSWERED,
 			Horde_Imap_Client::FLAG_FLAGGED,
+			Horde_Imap_Client::FLAG_FORWARDED,
 			Horde_Imap_Client::FLAG_DELETED,
 			Horde_Imap_Client::FLAG_DRAFT,
-			Horde_Imap_Client::FLAG_RECENT,
 			Horde_Imap_Client::FLAG_JUNK,
+			Horde_Imap_Client::FLAG_NOTJUNK,
 			Horde_Imap_Client::FLAG_MDNSENT,
+			Horde_Imap_Client::FLAG_RECENT,
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:735]: nope nope nope :)

File: lib/Service/Provisioning/ConfigMapper.php [hunk 5]
```diff
@@ -38,10 +38,10 @@ public function __construct(IConfig $config) {
 		$this->config = $config;
 	}
 
-	public function load(): ?Config {
+	public function load(?string $configKey = null): ?Config {
```
Comment (ChristophWurst) [inline lib/Service/Provisioning/ConfigMapper.php:41]: remnant of an older experiment?

**Review by ChristophWurst — COMMENTED** [review]
some early feedback

File: lib/Service/Provisioning/ConfigMapper.php (hunk 5 above)
Comment (miaulalala) [reply lib/Service/Provisioning/ConfigMapper.php:41]: Indeed :)

File: lib/Model/IMAPMessage.php (hunk 4 above)
Comment (miaulalala) [reply lib/Model/IMAPMessage.php:735]: That shouldn't even be there! :grimacing:

File: lib/Db/ProvisioningMapper.php (hunk 3 above)
Comment (miaulalala) [reply lib/Db/ProvisioningMapper.php]: It's in the Refactoring To Do ;)

**Comment by miaulalala** [comment]
Still to do:

- [ ] General Refactor to stop the Entity Creep (it's everywhere...)
- [ ] Refactor the ProvisioningMapper methods
- [ ] Tests
- [ ] Some work on the Vue Components, I don't like that they're pretty unresponsive when de-/activating a provisioning config
- [ ] Whatever else comes up during testing

File: lib/Db/MailAccount.php [hunk 6]
```diff
@@ -213,7 +216,7 @@ public function toJson() {
 			'imapSslMode' => $this->getInboundSslMode(),
 			'signature' => $this->getSignature(),
 			'editorMode' => $this->getEditorMode(),
-			'provisioned' => ($this->getProvisioned() === true),
+			'provisioningId' => ($this->getProvisioningId()),
```
Comment (ChristophWurst) [inline lib/Db/MailAccount.php:219]: nit

```suggestion
			'provisioningId' => $this->getProvisioningId(),
```

File: lib/Db/ProvisioningMapper.php [hunk 7]
```diff
@@ -42,30 +45,159 @@ public function getActive() : ?array {
...
-	public function getWildcardConfig() : ?array {
+	public function setActive(bool $active, Provisioning $provisioning) : void {
+		$qb = $this->db->getQueryBuilder();
+		$qb = $qb->update($this->getTableName())
+				->set('active', $qb->createNamedParameter($active, IQueryBuilder::PARAM_BOOL))
+				->where($qb->expr()->eq('provisioning_domain', $qb->createNamedParameter($provisioning->getProvisioningDomain())));
+		$qb->execute();
+	}
+
+	public function getAll() : array {
```
Comment (ChristophWurst) [inline lib/Db/ProvisioningMapper.php:75]: annotate the array type in a doc block

File: lib/Db/ProvisioningMapper.php [hunk 8]
```diff
@@ -42,30 +45,159 @@ public function getActive() : ?array {
...
+		$qb = $qb->select('*')
+				->from($this->getTableName())
+				->where($qb->expr()->eq('provisioning_domain', $qb->createNamedParameter(self::WILDCARD)));
+		$cursor = $qb->execute();
+		$result = $cursor->fetch();
+		$cursor->closeCursor();
+		if ($result === false) {
+			return null;
+		}
+		return Provisioning::fromRow($result);
```
Comment (ChristophWurst) [inline lib/Db/ProvisioningMapper.php:100]: use ``\OCP\AppFramework\Db\QBMapper::findOneQuery`` then you don't have to deal with the cursor and other repetitive code :)

File: lib/Exception/ValidationException.php [hunk 9]
```diff
@@ -21,29 +21,37 @@
...
 
-namespace OCA\mail\tests\Unit\Service\Provisioning;
+namespace OCA\Mail\Exception;
+
+use Exception;
+use OCP\AppFramework\Http;
 
-use OCA\Mail\Service\Provisioning\Config;
+class ValidationException extends Exception {
+	/** @var array */
```
Comment (ChristophWurst) [inline lib/Exception/ValidationException.php:33]:
```suggestion
	/** @var string[] */
```

for better static analysis

File: lib/Service/Provisioning/Manager.php [hunk 10]
```diff
@@ -155,17 +153,12 @@ private function updateAccount(IUser $user, MailAccount $account, Config $config
...
-			$config->setActive(false);
-			$this->configMapper->save($config);
-		}
+	public function deprovision(Provisioning $provisioning): void {
+		$this->mailAccountMapper->deleteProvisionedAccounts($provisioning->getId());
+		$this->provisioningMapper->setActive(false, $provisioning);
 	}
 
-	public function importConfig(array $data): Config {
+	public function importConfig(array $data): Provisioning {
```
Comment (ChristophWurst) [inline lib/Service/Provisioning/Manager.php:161]: the import is called from ``\OCA\Mail\Migration\MigrateProvisioningConfig::run`` and I think at this point we could even drop this old repair step as most instances will have upgraded.

**Review by ChristophWurst — COMMENTED** [review]
some early feedback

File: lib/Db/ProvisioningMapper.php (hunk 8 above)
Comment (miaulalala) [reply lib/Db/ProvisioningMapper.php:100]: :+1:  it's on the to do for tomorrow

File: lib/Service/Provisioning/Manager.php (hunk 10 above)
Comment (miaulalala) [reply lib/Service/Provisioning/Manager.php]: Deleted!

**Comment by miaulalala** [comment]
@kesselb would you have time for a review?

File: appinfo/routes.php [hunk 11]
```diff
@@ -244,6 +244,11 @@
 			'url' => '/api/settings/provisioning/{id}',
 			'verb' => 'POST'
 		],
+		[
+			'name' => 'settings#provision',
+			'url' => '/api/settings/provisioning/all',
+			'verb' => 'GET'
```
Comment (ChristophWurst) [inline appinfo/routes.php:250]: this is an action, this should be PUT or POST

Comment (miaulalala) [reply appinfo/routes.php:250]: Done, thanks!

**Comment by miaulalala** [comment]
Looks like the 7.4 tests are failing because of the insertOrUpdate Method that throws the UniqueConstraintViolation Exception. Any suggestions @ChristophWurst ?

File: lib/Http/Middleware/ProvisioningMiddleware.php [hunk 12]
```diff
@@ -64,12 +64,12 @@ public function beforeController($controller, $methodName) {
 			// Nothing to update
 			return;
 		}
-		$config = $this->provisioningManager->getConfig();
-		if ($config === null || !$config->isActive()) {
-			return;
+		$configs = $this->provisioningManager->getConfigs();
+		if (empty($configs)) {
```
Comment (kesselb) [inline lib/Http/Middleware/ProvisioningMiddleware.php:68]:
```suggestion
		if ($configs === []) {
```

File: lib/Db/ProvisioningMapper.php [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	/**
+	 * @return Provisioning[]
+	 */
+	public function getAll() : array {
+		$qb = $this->db->getQueryBuilder();
+		$qb = $qb->select('*')
+				->from($this->getTableName())
+				->orderBy('provisioning_domain', 'desc');
```
[the remark this answers is not in the document]
Comment (miaulalala) [reply lib/Db/ProvisioningMapper.php:56]: It's in the text for admin, but not the code. Good idea.

File: src/components/settings/AdminSettings.vue [hunk 14]
```diff
@@ -20,34 +20,209 @@
...
 			required: true,
 		},
 	},
+	data() {
+		return {
+			addNew : false,
+			formKey: Math.random(),
+			configs: this.provisioningSettings,
+			preview: {
+				active: true,
```
Comment (ChristophWurst) [inline src/components/settings/AdminSettings.vue:143]: `active` is dead

File: lib/Db/ProvisioningMapper.php [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+			$exception->setField('smtpHost', false);
+		}
+		if (!isset($data['smtpPort']) || (int)$data['smtpPort'] === 0) {
+			$exception->setField('smtpPort', false);
+		}
+		if (!isset($data['smtpSslMode']) || trim($data['smtpSslMode']) === '') {
+			$exception->setField('smtpSslMode', false);
+		}
+
+		if (empty($exception->getFields()) === false) {
```
[the remark this answers is not in the document]
Comment (miaulalala) [reply lib/Db/ProvisioningMapper.php:108]: yes - i still like a call to `empty` as it's more expressive.

File: lib/Model/IMAPMessage.php [hunk 16]
```diff
@@ -725,11 +725,12 @@ public function toDbMessage(int $mailboxId, MailAccount $account): Message {
 		$msg->setFlagMdnsent(in_array(Horde_Imap_Client::FLAG_MDNSENT, $flags, true));
 
 		$allowed = [
+			Horde_Imap_Client::FLAG_SEEN,
 			Horde_Imap_Client::FLAG_ANSWERED,
 			Horde_Imap_Client::FLAG_FLAGGED,
-			Horde_Imap_Client::FLAG_FORWARDED,
```
[the remark this answers is not in the document]
Comment (miaulalala) [reply lib/Model/IMAPMessage.php]: nope, but will disappear after rebasing

File: tests/Integration/Db/ProvisioningMapperTest.php [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+	private $db;
+
+	/** @var MockObject */
+	private $logger;
+
+	/** @var Alias */
+	private $alias;
+
+	/** @var [] */
+	public $data;
```
Comment (kesselb) [inline tests/Integration/Db/ProvisioningMapperTest.php:54]:
```suggestion
	public $data = [];
```

File: tests/Integration/Db/ProvisioningMapperTest.php [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+	/**
+	 * Initialize Mapper
+	 */
+	public function setup(): void {
+		parent::setUp();
+		$this->db = OC::$server->getDatabaseConnection();
+		$this->logger = $this->createMock(LoggerInterface::class);
+		$this->mapper = new ProvisioningMapper($this->db, $this->logger);
+		$this->data = [];
```
Comment (kesselb) [inline tests/Integration/Db/ProvisioningMapperTest.php:64]:
```suggestion
```

**Comment by ChristophWurst** [comment]
> [blocked ](/nextcloud/mail/labels/blocked)

Blocked because …?

**Comment by ChristophWurst** [comment]
> WIP: add multiple configs

Can we haz meaningful commit message? :)

File: src/components/settings/AdminSettings.vue [hunk 19]
```diff
@@ -20,34 +20,212 @@
...
+							"This setting only makes most sense if you use the same user back-end for your organization's Nextcloud and mail server."
+						)
+					}}
+				</p>
+			</article>
+		</div>
+		<h3>Provisioning Configurations</h3>
+		<p>
+			<button class="config-button icon-add" @click="addNew=true">
+				Add New Config
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply src/components/settings/AdminSettings.vue:97]: l10n is missing as well

**Review by ChristophWurst — APPROVED** [review]
Tested, fixed and now it seems to work as expected.

**Pull request merged**